### ISSUE TITLE
Uploading files to S3 with FileAdapter

### ISSUE THREAD

**Issue by ZacharyKhan** [parse-community/parse-server-example#119]
Im trying to store files to parse the way I have been via the iOS client, but I've set the backend up to store files in S3 (or at least I think I have). 

I am running my Parse-Server on AWS ElasticBeanstalk with mLab as my DB. 

I receive the following error when trying to upload a photo: `[Error]: Could not store file. (Code: 130, Version: 1.12.0)`

Here's a photo of my ElasticBeanstalk Configuration properties:
<img width="759" alt="screen shot 2016-03-20 at 6 05 54 pm" src="https://cloud.githubusercontent.com/assets/14354004/13908738/77f505aa-eec6-11e5-95b7-a1e5f16ee82a.png">

and my index.js is set up using the FileAdapter guide on GitHub

**Comment by jdzorz** [comment]
@ZacharyKhan in your index.js are you adding the s3 file properties  versus the elasticbeanstalk config properties. Since you are using express? 

Should look like this:

var S3Adapter = require('parse-server').S3Adapter;

var api = new ParseServer({
  databaseURI: databaseUri || 'mongodb://localhost:27017/dev',
  appId: process.env.APP_ID || 'APPLICATION_ID',
  masterKey: process.env.MASTER_KEY || 'MASTER_KEY',
  ...
  filesAdapter: new S3Adapter(
    "S3_ACCESS_KEY",
    "S3_SECRET_KEY",
    "S3_BUCKET",
    {directAccess: true}
  ),
  ...
});

Once you do that. restart your app servers

https://github.com/ParsePlatform/parse-server/wiki/Configuring-File-Adapters

**Comment by ZacharyKhan** [comment]
My _**index.js**_ is set up just like you have above, but replaced with my keys. 

When trying to store data to mongoDB and the file to S3, I run into the same error `[Error]: Could not store file. (Code: 130, Version: 1.12.0)`, and the entire post fails to save on MongoDB/S3. If I simply remove the adapter from **_index.js_**, files are properly stored within the MongoDB.

**Comment by jdzorz** [comment]
@ZacharyKhan can you go to \lib\routers\FilesRouter.js and in line 108 where you see this:

 next(new Parse.Error(Parse.Error.FILE_SAVE_ERROR, 'Could not store file.'));
change to:
next(new _node.Parse.Error(_node.Parse.Error.FILE_SAVE_ERROR, 'Could not store file.' + error));

this should give you any S3 errors. that might be occurring on that end. You'll have to restart EB app server once you make the change.

**Comment by ZacharyKhan** [comment]
Where exactly might I find the  `\lib\routers\FilesRouter.js`? There is no `lib` folder in my Parse-Server.

**Comment by jdzorz** [comment]
@ZacharyKhan  do you not see it under   /var/app/current/node_modules/parse-server/lib  ?

**Comment by ZacharyKhan** [comment]
I am running the Express version if that makes a difference? I dont see it

**Comment by jdzorz** [comment]
@ZacharyKhan are you viewing the files directly from your ec2 server on elastic beanstalk? You'll have to ssh into it. Once you connect then you'll see the files under /var/app/current/node_modules/parse-server

**Comment by xor22h** [comment]
It still has problems. For me issue was S3 storage in Frankfurt. Frankfurt doesn't support old signature (It request v4 signature). Seems like S3Adapter doesn't support that;

"The authorization mechanism you have provided is not supported. Please use AWS4-HMAC-SHA256"

**Comment by jeremypiednoel** [comment]
Did you make de policy on s3 ?
https://github.com/ParsePlatform/parse-server/wiki/Configuring-File-Adapters#set-up-your-bucket-and-permissions

**Comment by jayjac** [comment]
Same fix as xor22h : changing the bucket's region solved the issue. Frankfurt that was automatically selected for me doesn't support the old signature. I had to create a bucket in US Standard region to get the adapter working.

**Comment by jeremypiednoel** [comment]
Dublin bucket worked well for me

**Comment by abbasdawood** [comment]
Anyone able to figure this issue yet? I am still get the code 130 error, and I understand its got something to do with the content-type of the file. I am manually setting it though (to 'text/csv' as the file is a csv file in my case) and still getting the following response:

`Could not store file.InvalidParameterType: Expected params.ContentType to be a string`

**Comment by abbasdawood** [comment]
Solved this by changing my S3 Adapter require statement from 

`var S3Adapter = require('parse-server-s3-adapter')` 

to 

`var S3Adapter = require('parse-server').S3Adapter`

**Comment by kalyandechiraju** [comment]
Changing the region worked like magic!! Thank you. (I changed to Oregon)

**Comment by kesongxie** [comment]
How do you get access to the index.js in the elasticbeanstalk? I just set up the environment variable in the dashboard following the configuration [Here](https://github.com/ParsePlatform/parse-server/wiki/Configuring-File-Adapters#configuring-s3adapter ) I upload the the file as normal using this

```
let post = PFObject(className: "Post")
        post["description"] = "hello world"
        do{
            post["media"] = try PFFile(data: Data(contentsOf: url))
        }catch let error as NSError{
            print(error.localizedDescription)
        }
        post.saveInBackground { (success, error) in
            if success{
                print("video saved")
            }else{
                print(error?.localizedDescription)
            }
        }
 


```
But the media file is saved to my mLab like this, and my S3 bucket is currently empty
```
{
    "_id": "Cx24LadasKH8",
    "media": "f5f1071836544asf2420632722d569c9f_file.bin",
    "description": "hello world",
    "_created_at": {
        "$date": "2017-01-26T01:52:45.141Z"
    },
    "_updated_at": {
        "$date": "2017-01-26T01:52:45.141Z"
    }
}
```

Can I save file to S3 just using Parse SDK, or do I have to install Amazon iOS SDK and use S3TransferManager? Can anyone point me to the right direction. I'm new to Parse and Amazon Webservice

**Comment by jdzorz** [comment]
@kesongxie to connect to your elastic beanstalk you have to ssh into the ec2 instance. From there you can go to the index.js. You'll have to setup an s3 bucket with the permissions then try changing the configuration for the parse server to connect to it and save there.

**Comment by kesongxie** [comment]
@jdzorz Thanks. I figured it out! What I did is download the original parse-server-example.zip and modified and re-uploaded to the dashboard(I have to go inside the folder and zip all the items together instead zip the global folder). My S3 bucket along with other configurations were set up properly by following the S3Adapter article and the article for setting up parse beanstalk from Amazon. Now I can use the pretty much the same code with PFFile and saveInBackground to save my file to the S3 and created a reference field in my mLab as well.
